dss/util/security.py

```diff
@@ -0,0 +1,128 @@
+
+
+@functools.lru_cache(maxsize=32)
+def get_public_keys(openid_provider):
+    keys = requests.get(get_jwks_uri(openid_provider)).json()["keys"]
+    return {
+        key["kid"]: rsa.RSAPublicNumbers(
+            e=int.from_bytes(base64.urlsafe_b64decode(key["e"] + "==="), byteorder="big"),
+            n=int.from_bytes(base64.urlsafe_b64decode(key["n"] + "==="), byteorder="big")
+        ).public_key(backend=default_backend())
+        for key in keys
+    }
+
+
+def get_token_info(token):
+    verified_token = verify_jwt(token)
+    if Config.get_token_info_url() in verified_token['aud']:
+        token_request = session.get(Config.get_token_info_url(),
+                                    headers={'Authorization': 'Bearer {}'.format(token)},
+                                    timeout=5)
+        verified_token.update(token_request.json())
+        if verified_token.get('email_verified') is not True:
+            logger.info("Email not verified: {verified_token}")
+            raise DSSForbiddenException()
+    else:
+        # then token is a google service credential from a trusted google project.
+        verified_token['email'] = verified_token['sub']
+        if not (verified_token.get('scope') and verified_token.get('scopes')):
+            verified_token['scopes'] = []
+    return verified_token
+
+
+def verify_jwt(token: str) -> typing.Optional[typing.Mapping]:
+    try:
+        unverified_token = jwt.decode(token, verify=False)
+        issuer = unverified_token["iss"]
+        is_authorized_issuer(issuer)
+        public_keys = get_public_keys(issuer)
+        token_header = jwt.get_unverified_header(token)
+        if Config.deployment_stage() != 'prod':
+            logging.info(f"Token expiration is not checked in non production stages. Token: {token}")
+            tok = jwt.decode(token,
+                             key=public_keys[token_header["kid"]],
+                             issuer=issuer,
+                             audience=Config.get_audience(),
+                             algorithms=allowed_algorithms,
+                             options={'verify_exp': False})
+        else:
+            tok = jwt.decode(token,
+                             key=public_keys[token_header["kid"]],
+                             issuer=issuer,
+                             audience=Config.get_audience(),
+                             algorithms=allowed_algorithms,
+                             )
+    except jwt.PyJWTError as ex:  # type: ignore
+        logger.info(f"JWT failed to validate.", exc_info=True)
+        raise DSSException(401, 'Unauthorized', 'Authorization token is invalid') from ex
+    return tok
+
+
+def is_authorized_issuer(issuer: str) -> None:
+    """
+    Must be either "https://humancellatlas.auth0.com/" or service credential from a trusted google project.
+    :param issuer: str
+    """
+    if issuer != Config.get_openid_provider() and not any(issuer.endswith(f"@{p}")
```

Comment:
As per the above, I think the project names should not be conflated with service account ID domains (`@PROJECT_NAME.iam.gserviceaccount.com`).